tools/cluster_mgr.py

```diff
@@ -153,11 +149,180 @@ def main():
         update_id(n)
     print()
 
-    config = build_config(masters)
-    print(f'Pushing config...')
-    push_config(nodes, config)
+    config = build_config_from_list(masters)
+    print(f'Pushing config:\n{config}\n')
+    push_config(config)
     print()
 
 
+def build_config_from_existing(args):
+    def list_to_dict(l):
+        return {l[i]: l[i + 1] for i in range(0, len(l), 2)}
+
+    def build_node(node_list):
+        d = list_to_dict(node_list)
+        return {
+            "id": d["id"],
+            "ip": d["endpoint"],
+            "port": d["port"]
+        }
+
+    def build_slots(slot_list):
+        slots = []
+        for i in range(0, len(slot_list), 2):
+            slots.append(
+                {
+                    "start": slot_list[i],
+                    "end": slot_list[i+1]
+                }
+            )
+        return slots
+
+    client = redis.Redis(decode_responses=True,
+                         host="localhost", port=args.first_port)
+    existing = client.execute_command("cluster", "shards")
+    config = []
+    for shard_list in existing:
+        shard = list_to_dict(shard_list)
+        config.append({
+            "slot_ranges": build_slots(shard["slots"]),
+            "master": build_node(shard["nodes"][0]),
+            "replicas": [build_node(replica) for replica in shard["nodes"][1::]]
+        })
+    return config
+
+
+def move(args):
+    config = build_config_from_existing(args)
+
+    new_owner = None
+    for shard in config:
+        if shard["master"]["port"] == args.target_port:
+            new_owner = shard
+            break
+    else:
+        print(
+            f"Can't find master with port {args.target_port} (hint: use flag --target_port).")
+        exit(-1)
+
+    def remove_slot(slot, from_range, from_shard):
+        if from_range["start"] == slot:
+            from_range["start"] += 1
+            if from_range["start"] > from_range["end"]:
+                from_shard["slot_ranges"].remove(from_range)
+        elif from_range["end"] == slot:
+            from_range["end"] -= 1
+            if from_range["start"] > from_range["end"]:
+                from_shard["slot_ranges"].remove(from_range)
+        else:
+            assert slot > from_range["start"] and slot < from_range["end"], f'{slot} {from_range["start"]} {from_range["end"]}'
+            from_shard["slot_ranges"].append(
+                {"start": slot + 1, "end": from_range["end"]})
+            from_range["end"] = slot - 1
```

Comment:
Quite hard to read 🔎 

Actually if you will have more operations it would make sense to use dedicated objects with properties here. Json parsing will require some additional magic, but you can already verify schema correctness there as well